### PR TITLE
lang: programdata_address field for Program account

### PR DESCRIPTION
**Pull request by paul-schaaf**

### CODE REVIEW

File: lang/src/program.rs [hunk 1]
```diff
@@ -1,31 +1,44 @@
...
 use solana_program::account_info::AccountInfo;
+use solana_program::bpf_loader_upgradeable::{self, UpgradeableLoaderState};
 use solana_program::instruction::AccountMeta;
 use solana_program::program_error::ProgramError;
 use solana_program::pubkey::Pubkey;
 use std::fmt;
 use std::ops::Deref;
 
+// TODO: can we remove _account?
+
```
Comment (paul-schaaf) [inline lang/src/program.rs:12]: @armaniferrante ^

is there any use to this field, considering that it's going to be serialized to an empty struct I think all(?) of the time

File: lang/src/program.rs [hunk 2]
```diff
@@ -1,31 +1,44 @@
...
 use crate::*;
 use solana_program::account_info::AccountInfo;
+use solana_program::bpf_loader_upgradeable::{self, UpgradeableLoaderState};
 use solana_program::instruction::AccountMeta;
 use solana_program::program_error::ProgramError;
 use solana_program::pubkey::Pubkey;
 use std::fmt;
 use std::ops::Deref;
 
+// TODO: can we remove _account?
```
Comment (armaniferrante) [inline lang/src/program.rs:11]:
```suggestion
```

File: lang/src/program.rs (hunk 1 above)
Comment (armaniferrante) [inline lang/src/program.rs:12]:
```suggestion
```

File: lang/src/program.rs [hunk 3]
```diff
@@ -37,9 +50,44 @@ impl<'a, T: Id + AccountDeserialize + Clone> Program<'a, T> {
...
+                | UpgradeableLoaderState::Buffer {
+                    authority_address: _,
+                }
+                | UpgradeableLoaderState::ProgramData {
+                    slot: _,
+                    upgrade_authority_address: _,
+                } => {
+                    // unreachable because check above already
+                    // ensures that program is executable
+                    // and therefore a program account
```
Comment (armaniferrante) [inline lang/src/program.rs:69]:
```suggestion
                    // and therefore a program account.
```

File: lang/src/program.rs [hunk 4]
```diff
@@ -37,9 +50,44 @@ impl<'a, T: Id + AccountDeserialize + Clone> Program<'a, T> {
...
+            match upgradable_loader_state {
+                UpgradeableLoaderState::Uninitialized
+                | UpgradeableLoaderState::Buffer {
+                    authority_address: _,
+                }
+                | UpgradeableLoaderState::ProgramData {
+                    slot: _,
+                    upgrade_authority_address: _,
+                } => {
+                    // unreachable because check above already
```
Comment (armaniferrante) [inline lang/src/program.rs:67]:
```suggestion
                    // Unreachable because check above already
```

File: lang/src/program.rs (hunk 1 above)
Comment (paul-schaaf) [reply lang/src/program.rs]: yes that sounds better to me. https://github.com/project-serum/anchor/issues/1140

**Pull request merged**